src/pytest_workflow/workflow.py

```diff
@@ -17,54 +17,171 @@
+        self.wait_timeout_secs = None
+        self.wait_interval_secs = 0.01
+        self.wait_counter = 0
+
+    def start(self):
+        """Runs the workflow in a subprocess in the background.
+        To make sure the workflow is finished use the `.wait()` method"""
+        # The lock ensures that the workflow is started only once, even if it
+        # is started from multiple threads.
+        with self.start_lock:
+            if self._popen is None:
+                sub_procces_args = shlex.split(self.command)
+                self._popen = subprocess.Popen(  # nosec: Shell is not enabled.
+                    sub_procces_args, stdout=subprocess.PIPE,
+                    stderr=subprocess.PIPE, cwd=str(self.cwd))
+            else:
+                raise ValueError("Workflows can only be started once")
 
     def run(self):
-        sub_procces_args = shlex.split(self.command)
-        self._proc_out = subprocess.run(  # nosec: Shell is not enabled.
-            sub_procces_args, stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE, cwd=str(self.cwd))
+        """Runs the workflow and blocks until it is finished"""
+        self.start()
+        self.wait()
 
     def stdout_to_file(self) -> Path:
+        self.wait()
         return bytes_to_file(self.stdout, self.cwd / Path("log.out"))
 
     def stderr_to_file(self) -> Path:
-        return bytes_to_file(self.stdout, self.cwd / Path("log.err"))
+        self.wait()
+        return bytes_to_file(self.stderr, self.cwd / Path("log.err"))
+
+    def wait(self):
+        """Waits for the workflow to complete"""
+        # Lock the wait step. Only one waiter is allowed here to wait for
+        # the workflow to complete and write the stderr.
+        # A popen.stderr is a buffered reader and can only be read once.
+        # Once self._stderr and self._stdout are written the lock can be
+        # released
+        with self.wait_lock:
+            while self._popen is None:
+                # This piece of code checks if a workflow has started yet. If
+                # it has not, it waits. A counter is implemented here because:
+                # 1. Incrementing a counter is much faster than checking system
+                #    time
+                # 2. The counter is linked to the self object. This means we
+                #    wait only once for the workflow to start. All consecutive
+                #    wait commands will fail instantly. Having all of these
+                #    wait as well would be a waste of time.
+                if self.wait_timeout_secs is not None and self.wait_counter > (
+                        self.wait_timeout_secs / self.wait_interval_secs):
+                    raise ValueError(
+                        "Waiting on a workflow that has not started within the"
+                        " last {0} seconds".format(self.wait_timeout_secs))
+                time.sleep(self.wait_interval_secs)
+                self.wait_counter += 1
```

Comment:
wouldn't it be easier to simply count on the number of seconds? i.e

```
if self.wait_time_seconds > self.max_wait_seconds:
    raise ValueError
time.sleep(self.wait_interval_secs)
self.wait_time_secs += self.wait_internval_secs
```

Reply:
But that's floating point calculation. We should use integers for premature optimization overengineering concerns!!!1!1!!1!